File: android/app/src/main/java/com/retailcalcpro/app/util/UsageLimiter.kt
package com.retailcalcpro.app.util

import android.content.Context
import android.content.SharedPreferences
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class UsageLimiter private constructor(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("usage_limiter", Context.MODE_PRIVATE)

    companion object {
        private const val DAILY_LIMIT = 10
        private const val KEY_COUNT = "dailyUsageCount"
        private const val KEY_DATE = "dailyUsageDate"

        @Volatile
        private var INSTANCE: UsageLimiter? = null

        fun getInstance(context: Context): UsageLimiter {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: UsageLimiter(context.applicationContext).also { INSTANCE = it }
            }
        }
    }

    val isLimitReached: Boolean
        get() {
            if (PurchaseManager.getInstance().isProUnlocked) return false
            loadTodayUsage()
            return todayUsageCount >= DAILY_LIMIT
        }

    val remainingCount: Int
        get() {
            if (PurchaseManager.getInstance().isProUnlocked) return 999
            loadTodayUsage()
            return maxOf(0, DAILY_LIMIT - todayUsageCount)
        }

    private var todayUsageCount: Int = 0

    init {
        loadTodayUsage()
    }

    fun recordUsage(): Boolean {
        if (PurchaseManager.getInstance().isProUnlocked) return true

        loadTodayUsage()

        if (todayUsageCount >= DAILY_LIMIT) {
            return false
        }

        todayUsageCount++
        saveTodayUsage()
        return true
    }

    fun canUse(): Boolean {
        if (PurchaseManager.getInstance().isProUnlocked) return true
        loadTodayUsage()
        return todayUsageCount < DAILY_LIMIT
    }

    private fun loadTodayUsage() {
        val savedDate = prefs.getString(KEY_DATE, "") ?: ""
        val today = todayString()

        if (savedDate == today) {
            todayUsageCount = prefs.getInt(KEY_COUNT, 0)
        } else {
            todayUsageCount = 0
            prefs.edit()
                .putString(KEY_DATE, today)
                .putInt(KEY_COUNT, 0)
                .apply()
        }
    }

    private fun saveTodayUsage() {
        prefs.edit()
            .putString(KEY_DATE, todayString())
            .putInt(KEY_COUNT, todayUsageCount)
            .apply()
    }

    private fun todayString(): String {
        val formatter = SimpleDateFormat("yyyy-MM-dd", Locale.JAPAN)
        return formatter.format(Date())
    }
}

File: RetailCalcPro/RetailCalcPro/Utils/UsageLimiter.swift
import Foundation

/// 1日の利用回数を管理するクラス
/// 「C（クリア）」で ¥0 に戻ったタイミングを「1回の利用」としてカウント
@MainActor
@Observable
final class UsageLimiter {
    static let shared = UsageLimiter()

    private static let dailyLimit = 10
    private static let usageCountKey = "dailyUsageCount"
    private static let usageDateKey = "dailyUsageDate"

    private(set) var todayUsageCount: Int = 0
    var showPaywall: Bool = false

    var isLimitReached: Bool {
        if PurchaseManager.shared.isProUnlocked { return false }
        return todayUsageCount >= Self.dailyLimit
    }

    var remainingCount: Int {
        if PurchaseManager.shared.isProUnlocked { return 999 }
        return max(0, Self.dailyLimit - todayUsageCount)
    }

    private init() {
        loadTodayUsage()
    }

    /// 利用回数をインクリメント
    @discardableResult
    func recordUsage() -> Bool {
        if PurchaseManager.shared.isProUnlocked { return true }

        loadTodayUsage()

        if todayUsageCount >= Self.dailyLimit {
            showPaywall = true
            return false
        }

        todayUsageCount += 1
        saveTodayUsage()
        return true
    }

    /// 入力開始時に制限をチェック
    func canUse() -> Bool {
        if PurchaseManager.shared.isProUnlocked { return true }

        loadTodayUsage()
        if todayUsageCount >= Self.dailyLimit {
            showPaywall = true
            return false
        }
        return true
    }

    private func loadTodayUsage() {
        let defaults = UserDefaults.standard
        let savedDate = defaults.string(forKey: Self.usageDateKey) ?? ""
        let today = Self.todayString()

        if savedDate == today {
            todayUsageCount = defaults.integer(forKey: Self.usageCountKey)
        } else {
            todayUsageCount = 0
            defaults.set(today, forKey: Self.usageDateKey)
            defaults.set(0, forKey: Self.usageCountKey)
        }
    }

    private func saveTodayUsage() {
        let defaults = UserDefaults.standard
        defaults.set(Self.todayString(), forKey: Self.usageDateKey)
        defaults.set(todayUsageCount, forKey: Self.usageCountKey)
    }

    private static func todayString() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.locale = Locale(identifier: "ja_JP")
        return formatter.string(from: Date())
    }
}
